### DISCUSSION TITLE
Is it supported to make rust crate at root and npm project at child directory?

### DISCUSSION THREAD

**Discussion by anatawa12** [tauri-apps/tauri#8890] [Q&A] [answered]
I'm going to develop an gui app with tauri and I already have some cli version in rust so I want to make directory structure like the following:

```
  - repository root (cargo workspace)
     +- common library (rust)
     +- cli version (rust)
     `- gui rust code (rust)
         `- gui web code (npm)
```

However, in the tauri documentation, there is no documentation for such a use case.
(In [Integrate into Existing Project](https://tauri.app/v1/guides/getting-started/setup/integrate#running-your-app), integrating to existing web app is described but integrating to existing rust project is not.)

Of course, I can make like following and add the rust project to the repository root but I don't like this structure.

```
  - repository root (cargo workspace)
     +- common library (rust)
     +- cli version (rust)
     `- gui web code (npm)
         `- gui rust code (rust)
```

**Comment by FabianLars** [comment, marked as answer]
Yes it is, though it may requires to use `cargo tauri` (via `cargo install tauri-cli`) instead of the npm based cli, unless you're okay with also having a package.json file at the repo root.

`cargo tauri` will actively search for the tauri config file in the current working dir (including subfolders) so the exact repo structure doesn't really matter as long as you configure beforeDev|BuildCommand and devPath/distDir accordingly.

**Comment by RenaudDenis** [comment]
I much prefer this directory structure too.

However, after I install the tauri cli and move the files accordingly, `cargo tauri dev` seems to expect my `package.json` to be in the _parent_ directory of my project instead of the `web\` subdirectory.

So, if my project structure is as follows:

```
/sandbox/my-project — The base directory of my project and where my cargo.toml is
/sandbox/my-project/web — The web part of my project, containing the package.json
```

When I run `cargo tauri dev` in the base directory, it's looking for the package.json in `/sandbox` instead of `/sandbox/my-project/web`.

Here is a part of my `tauri.conf.json`:

```
  "build": {
    "beforeDevCommand": "npm run dev",
    "beforeBuildCommand": "npm run build",
    "devPath": "http://localhost:1420",
    "distDir": "dist/"
  },

```

Do you see what I am doing wrong?

**Reply by FabianLars** [reply]
you can specify the cwd for beforeDev|BuildCommand yourself. Something like this
```json
"beforeDevCommand": {
  "script": "npm run dev",
  "cwd": "web/",
```

**Reply by RenaudDenis** [reply]
Great! It works, many thanks
